File: packages/expo-papermint/android/src/main/java/expo/modules/papermint/ExpoPapermintModule.kt
package expo.modules.papermint

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException

class ExpoPapermintModule : Module() {
  companion object {
    private var isLibraryLoaded = false

    init {
      try {
        System.loadLibrary("papermint_mobile")
        isLibraryLoaded = true
      } catch (e: Throwable) {
        isLibraryLoaded = false
      }
    }
  }

  private external fun nativeCompileTicket(json: String, dialect: Int): ByteArray?
  private external fun nativeRenderSvg(json: String): String?
  private external fun nativeRenderHtml(json: String): String?
  private external fun nativeCompileLabel(json: String, dialect: Int): ByteArray?
  private external fun nativeRenderLabelSvg(json: String): String?

  override fun definition() = ModuleDefinition {
    Name("ExpoPapermint")

    Function("hello") {
      "Hello world! 👋"
    }

    Function("compileTicket") { json: String, dialect: String ->
      if (!isLibraryLoaded) {
        throw CodedException("ERR_PAPERMINT_NATIVE", "libpapermint_mobile.so is not loaded in this runtime", null)
      }
      val dialectCode = if (dialect.equals("star", ignoreCase = true)) 1 else 0
      val bytes = nativeCompileTicket(json, dialectCode)
        ?: throw CodedException("ERR_COMPILE_FAILED", "Failed to compile receipt ticket (invalid JSON payload or layout)", null)
      bytes
    }

    Function("renderSvg") { json: String ->
      if (!isLibraryLoaded) {
        throw CodedException("ERR_PAPERMINT_NATIVE", "libpapermint_mobile.so is not loaded in this runtime", null)
      }
      val svg = nativeRenderSvg(json)
        ?: throw CodedException("ERR_RENDER_SVG_FAILED", "Failed to render receipt SVG preview", null)
      svg
    }

    Function("renderHtml") { json: String ->
      if (!isLibraryLoaded) {
        throw CodedException("ERR_PAPERMINT_NATIVE", "libpapermint_mobile.so is not loaded in this runtime", null)
      }
      val html = nativeRenderHtml(json)
        ?: throw CodedException("ERR_RENDER_HTML_FAILED", "Failed to render receipt HTML preview", null)
      html
    }

    Function("compileLabel") { json: String, dialect: String ->
      if (!isLibraryLoaded) {
        throw CodedException("ERR_PAPERMINT_NATIVE", "libpapermint_mobile.so is not loaded in this runtime", null)
      }
      val dialectCode = if (dialect.equals("zpl", ignoreCase = true)) 1 else 0
      val bytes = nativeCompileLabel(json, dialectCode)
        ?: throw CodedException("ERR_COMPILE_LABEL_FAILED", "Failed to compile label (invalid JSON payload or layout)", null)
      bytes
    }

    Function("renderLabelSvg") { json: String ->
      if (!isLibraryLoaded) {
        throw CodedException("ERR_PAPERMINT_NATIVE", "libpapermint_mobile.so is not loaded in this runtime", null)
      }
      val svg = nativeRenderLabelSvg(json)
        ?: throw CodedException("ERR_RENDER_LABEL_SVG_FAILED", "Failed to render label SVG preview", null)
      svg
    }
  }
}

File: packages/expo-papermint/ios/ExpoPapermintModule.swift
import ExpoModulesCore
import Foundation

@_silgen_name("papermint_compile_json")
private func papermint_compile_json(
  _ jsonStr: UnsafePointer<CChar>?,
  _ dialect: UInt8,
  _ outLen: UnsafeMutablePointer<Int>?
) -> UnsafeMutablePointer<UInt8>?

@_silgen_name("papermint_bytes_free")
private func papermint_bytes_free(
  _ ptr: UnsafeMutablePointer<UInt8>?,
  _ len: Int
)

@_silgen_name("papermint_compile_json_svg")
private func papermint_compile_json_svg(
  _ jsonStr: UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>?

@_silgen_name("papermint_compile_json_html")
private func papermint_compile_json_html(
  _ jsonStr: UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>?

@_silgen_name("papermint_string_free")
private func papermint_string_free(
  _ ptr: UnsafeMutablePointer<CChar>?
)

public class ExpoPapermintModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoPapermint")

    Function("hello") {
      return "Hello world! 👋"
    }

    Function("compileTicket") { (json: String, dialect: String) -> Data in
      let dialectCode: UInt8 = (dialect.caseInsensitiveCompare("star") == .orderedSame) ? 1 : 0
      var outLen: Int = 0

      guard let utf8 = json.cString(using: .utf8) else {
        throw Exceptions.InvalidArgument("JSON ticket payload could not be encoded as UTF-8")
      }

      guard let ptr = papermint_compile_json(utf8, dialectCode, &outLen), outLen > 0 else {
        throw Exceptions.Fault("Failed to compile receipt ticket with Papermint engine (invalid layout or syntax)")
      }

      // Return Data wrapping the Rust-allocated byte buffer without copying.
      // Automatically freed via custom deallocator when JavaScript garbage collects the buffer.
      return Data(bytesNoCopy: ptr, count: outLen, deallocator: .custom { p, len in
        papermint_bytes_free(p.assumingMemoryBound(to: UInt8.self), len)
      })
    }

    Function("renderSvg") { (json: String) -> String in
      guard let utf8 = json.cString(using: .utf8) else {
        throw Exceptions.InvalidArgument("JSON ticket payload could not be encoded as UTF-8")
      }

      guard let ptr = papermint_compile_json_svg(utf8) else {
        throw Exceptions.Fault("Failed to render receipt SVG preview")
      }

      let result = String(cString: ptr)
      papermint_string_free(ptr)
      return result
    }

    Function("renderHtml") { (json: String) -> String in
      guard let utf8 = json.cString(using: .utf8) else {
        throw Exceptions.InvalidArgument("JSON ticket payload could not be encoded as UTF-8")
      }

      guard let ptr = papermint_compile_json_html(utf8) else {
        throw Exceptions.Fault("Failed to render receipt HTML preview")
      }

      let result = String(cString: ptr)
      papermint_string_free(ptr)
      return result
    }
  }
}

